src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncRewriter.AsyncIteratorRewriter.cs

```diff
@@ -161,50 +150,58 @@ protected override BoundStatement GenerateStateMachineCreation(LocalSymbol state
+                //  if (State == StateMachineStates.FinishedStateMachine)
+                //  {
+                //      return default(ValueTask<bool>)
+                //  }
+                //  _valueOrEndPromise.Reset();
+                //  var inst = this;
+                //  _builder.Start(ref inst);
+                //  return new ValueTask<bool>(this, _valueOrEndPromise.Version);
 
                 NamedTypeSymbol IAsyncEnumeratorOfElementType =
                     F.WellKnownType(WellKnownType.System_Collections_Generic_IAsyncEnumerator_T)
                     .Construct(_currentField.Type);
 
-                MethodSymbol IAsyncEnumerableOfElementType_WaitForNextAsync =
-                    F.WellKnownMethod(WellKnownMember.System_Collections_Generic_IAsyncEnumerator_T__WaitForNextAsync)
+                MethodSymbol IAsyncEnumerableOfElementType_MoveNextAsync =
+                    F.WellKnownMethod(WellKnownMember.System_Collections_Generic_IAsyncEnumerator_T__MoveNextAsync)
                     .AsMember(IAsyncEnumeratorOfElementType);
 
                 // The implementation doesn't depend on the method body of the iterator method.
-                OpenMethodImplementation( IAsyncEnumerableOfElementType_WaitForNextAsync, hasMethodBodyDependency: false);
-
-                BoundFieldAccess promiseIsActiveField = F.Field(F.This(), _promiseIsActiveField);
+                OpenMethodImplementation(IAsyncEnumerableOfElementType_MoveNextAsync, hasMethodBodyDependency: false);
 
                 var ifFinished = F.If(
                     // if (State == StateMachineStates.FinishedStateMachine)
                     F.IntEqual(F.Field(F.This(), stateField), F.Literal(StateMachineStates.FinishedStateMachine)),
-                    thenClause: F.Return(F.Default(IAsyncEnumerableOfElementType_WaitForNextAsync.ReturnType))); // return default(ValueTask<bool>)
+                    // return default(ValueTask<bool>)
+                    thenClause: F.Return(F.Default(IAsyncEnumerableOfElementType_MoveNextAsync.ReturnType)));
 
-                var ifNotRunningOrNotStarted = F.If(
-                    // if (!this._promiseIsActive || this.State == StateMachineStates.NotStartedStateMachine)
-                    F.Binary(BinaryOperatorKind.LogicalOr, promiseIsActiveField.Type,
-                        F.Not(promiseIsActiveField),
-                        F.IntEqual(F.Field(F.This(), stateField), F.Literal(StateMachineStates.NotStartedStateMachine))),
-                    thenClause: GenerateCallStart()); // var inst = this; this._builder.Start(ref inst);
+                // _promiseOfValueOrEnd.Reset();
+                BoundFieldAccess promiseField = F.Field(F.This(), _promiseOfValueOrEndField);
+                var resetMethod = (MethodSymbol)F.WellKnownMethod(WellKnownMember.System_Threading_Tasks_ManualResetValueTaskSourceLogic_T__Reset, isOptional: true)
+                    .SymbolAsMember((NamedTypeSymbol)_promiseOfValueOrEndField.Type);
```

Comment:
It can, but we checked all the required members before getting here.
`VerifyPresenceOfRequiredAPIs` is used to skip async-rewriting if any of the required APIs is missing.
Also, this is covered by test `MissingTypeAndMembers_ManualResetValueTaskSourceLogic`.

---
In reply to: [223523143](https://github.com/dotnet/roslyn/pull/30280#discussion_r223523143) [](ancestors = 223523143)